source/base/data_out_base.cc

```diff
@@ -9169,7 +9225,29 @@ XDMFEntry::XDMFEntry(const std::string & mesh_filename,
   , num_cells(cells)
   , dimension(dim)
   , space_dimension(spacedim)
-{}
+  , cell_type(cell_type_)
+{
+  if (cell_type == ReferenceCells::Invalid)
+    {
+      switch (dimension)
+        {
+          case 0:
+            cell_type = ReferenceCells::get_hypercube<0>();
+            break;
+          case 1:
+            cell_type = ReferenceCells::get_hypercube<1>();
+            break;
+          case 2:
+            cell_type = ReferenceCells::get_hypercube<2>();
+            break;
+          case 3:
+            cell_type = ReferenceCells::get_hypercube<3>();
+            break;
+          default:
+            AssertThrow(false, ExcMessage("Invalid dimension"));
+        }
+    }
```

Comment:
Could we move this logic into a helper function and call it directly in the other ctor?